// geomonster/index.js
var GeoMonster = function() {
	var self = this;

	/**
	 * This is the bounding box for which players and monsters need to stay and spawn
	 * Michigan: Lansing, Ann Arbor, Detroit, Jackson, Troy
	 */
	var bounds = {
		min: {
			latitude: 42.0000,
			longitude: -85.0000,
		},
		max: {
			latitude: 43.0000,
			longitude: -83.0000,
		}
	};

	var monster_types = {
		0: { name: 'Special 1', wild: false },
		1: { name: 'Special 2', wild: false },
		2: { name: 'Special 3', wild: false },

		3: { name: 'Raccoon', wild: true },
		4: { name: 'Salamander', wild: true },
		5: { name: 'Cobra', wild: true },
		6: { name: 'Bear', wild: true },
		7: { name: 'Tiger', wild: true },
		8: { name: 'Jiraffe', wild: true },
		9: { name: 'Rapist', wild: true },
	};

	var spawnable_monster_ids = [];

	for (var index in monster_types) {
		var current_monster = monster_types[index];
		if (current_monster.wild) {
			spawnable_monster_ids.push(index);
		}
	}

	var monsters = [];

	self.initializeMonsterPopulation = function(count) {
		for (var i = 0; i < count; i++) {
			monsters.push({
				coords: self.getRandomLocationWithinBounds(),
				type: self.getIdForRandomSpawnableMonster(),
			});
		}
	};

	self.getAllMonsters = function() {
		return monsters;
	};

	self.getMonsterTypes = function() {
		return monster_types;
	};

	self.getIdForRandomSpawnableMonster = function() {
		return spawnable_monster_ids[Math.floor(Math.random()*spawnable_monster_ids.length)];
	};

	/**
	 * Returns a location within our bounding box. It's an object with a lat/lon pair, that's it.
	 */
	self.getRandomLocationWithinBounds = function() {
		return {
			latitude: (Math.random() * (bounds.max.latitude - bounds.min.latitude)) + bounds.min.latitude,
			longitude: (Math.random() * (bounds.max.longitude - bounds.min.longitude)) + bounds.min.longitude,
		};
	};

	/**
	 * Checks to see if the provided location is within our bounding box.
	 */
	self.checkIfLocationWithinBounds = function(c) {
		return c.latitude >= bounds.min.latitude && c.latitude <= bounds.max.latitude && c.longitude >= bounds.min.longitude && c.longitude <= bounds.max.longitude;
	};
};

module.exports = new GeoMonster();
