// geomonster/index.js
var GeoMonster = function() {
	var self = this;

	var monster_move_wait_seconds = 4;

	var monster_max_move = {
		lat: 0.0001,
		lon: 0.0001,
	};

	/**
	 * This is the bounding box for which players and monsters need to stay and spawn
	 * Michigan: Lansing, Ann Arbor, Detroit, Jackson, Troy
	 */
	var bounds = {
		min: {
			lat: 42.0000,
			lon: -85.0000,
		},
		max: {
			lat: 43.0000,
			lon: -83.0000,
		}
	};

	var monster_types = {
		0: { name: 'Special 1', wild: false },
		1: { name: 'Special 2', wild: false },
		2: { name: 'Special 3', wild: false },

		3: { name: 'Raccoon', wild: true },
		4: { name: 'Salamander', wild: true },
		5: { name: 'Cobra', wild: true },
		6: { name: 'Bear', wild: true },
		7: { name: 'Tiger', wild: true },
		8: { name: 'Jiraffe', wild: true },
		9: { name: 'Rapist', wild: true },
	};

	var spawnable_monster_ids = [];

	for (var index in monster_types) {
		var current_monster = monster_types[index];
		if (current_monster.wild) {
			spawnable_monster_ids.push(index);
		}
	}

	var monsters = [];

	var monster_move_timer = null;

	self.initializeMonsterPopulation = function(count) {
		for (var i = 0; i < count; i++) {
			monsters.push({
				coords: self.getRandomLocationWithinBounds(),
				type: self.getIdForRandomSpawnableMonster(),
			});
		}
		return self;
	};

	self.initializeMonsterMovement = function() {
		monster_move_timer = setInterval(self.moveMonsters, 1000 * monster_move_wait_seconds);
		return self;
	};

	self.moveMonsters = function() {
		var number_of_monsters = monsters.length;
		for (var i = 0; i < number_of_monsters; i++) {
			new_coords = {
				lat: monsters[i].coords.lat + Math.random()*monster_max_move.lat - monster_max_move.lat/2,
				lon: monsters[i].coords.lon + Math.random()*monster_max_move.lon - monster_max_move.lon/2,
			};
			if (self.checkIfLocationWithinBounds(new_coords)) {
				monsters[i].coords = new_coords;
				// emit monster move?
			}
		}
		console.log(monsters);
	};

	self.killMonsterMovement = function() {
		clearTimeout(monster_move_timer);
	};

	self.getAllMonsters = function() {
		return monsters;
	};

	self.getMonsterTypes = function() {
		return monster_types;
	};

	self.getIdForRandomSpawnableMonster = function() {
		return spawnable_monster_ids[Math.floor(Math.random()*spawnable_monster_ids.length)];
	};

	/**
	 * Returns a location within our bounding box. It's an object with a lat/lon pair, that's it.
	 */
	self.getRandomLocationWithinBounds = function() {
		return {
			lat: (Math.random() * (bounds.max.lat - bounds.min.lat)) + bounds.min.lat,
			lon: (Math.random() * (bounds.max.lon - bounds.min.lon)) + bounds.min.lon,
		};
	};

	/**
	 * Checks to see if the provided location is within our bounding box.
	 */
	self.checkIfLocationWithinBounds = function(c) {
		return c.lat >= bounds.min.lat && c.lat <= bounds.max.lat && c.lon >= bounds.min.lon && c.lon <= bounds.max.lon;
	};
};

module.exports = new GeoMonster();
